File: backend/app/api/v1/records.py
"""The records table and one record: /api/v1/projects/{pid}/records (guide 10).

Any member may read records; what they may see of other people's decisions is settled in
the screening phase. Every query is scoped to the verified project.
"""

import uuid
from typing import Annotated

from fastapi import APIRouter, Query

from app.api.deps import RecordsDep, ViewerAccess
from app.api.responses import PROJECT
from app.models import FullTextStatus, TitleAbstractStatus
from app.schemas.problem import problem_content
from app.schemas.records import RecordDetail, RecordFacets, RecordPage, Sort
from app.services.pagination import DEFAULT_LIMIT, MAX_LIMIT

router = APIRouter(prefix="/projects/{pid}", tags=["records"])

Search = Annotated[
    str,
    Query(
        max_length=500,
        description='Words, "exact phrases", -exclusions, author:, journal:, keyword:, '
        "year:2018..2024, a DOI or a PubMed id (guide 8.9)",
    ),
]


@router.get("/records", responses={**PROJECT, 400: problem_content()})
async def list_records(
    access: ViewerAccess,
    records: RecordsDep,
    q: Search = "",
    status: TitleAbstractStatus | None = None,
    full_text: FullTextStatus | None = None,
    batch: uuid.UUID | None = None,
    duplicates: bool = False,
    sort: Sort = "added",
    cursor: Annotated[str | None, Query(max_length=512)] = None,
    limit: Annotated[int, Query(ge=1, le=MAX_LIMIT)] = DEFAULT_LIMIT,
) -> RecordPage:
    """The records table: filtered, sorted and paged."""
    return await records.page(
        access,
        search=q,
        ta=status,
        ft=full_text,
        batch_id=batch,
        duplicates=duplicates,
        sort=sort,
        cursor=cursor,
        limit=limit,
    )


@router.get("/records/facets", responses=PROJECT)
async def record_facets(access: ViewerAccess, records: RecordsDep) -> RecordFacets:
    """How many records sit behind each filter."""
    return await records.facets(access)


@router.get("/records/{rid}", responses=PROJECT)
async def get_record(rid: uuid.UUID, access: ViewerAccess, records: RecordsDep) -> RecordDetail:
    return await records.detail(access, rid)

File: frontend/src/api/records.ts
import { infiniteQueryOptions, queryOptions } from "@tanstack/react-query";

import { api, unwrap } from "@/api/client";
import type { components } from "@/api/schema";

export type Record = components["schemas"]["RecordOut"];
export type RecordDetail = components["schemas"]["RecordDetail"];
export type RecordPage = components["schemas"]["RecordPage"];
export type RecordFacets = components["schemas"]["RecordFacets"];
export type TitleAbstractStatus = components["schemas"]["TitleAbstractStatus"];
export type FullTextStatus = components["schemas"]["FullTextStatus"];
export type Sort = "added" | "oldest" | "year" | "year_asc" | "title" | "relevance";

export interface RecordQuery {
  q?: string;
  status?: TitleAbstractStatus;
  full_text?: FullTextStatus;
  batch?: string;
  duplicates?: boolean;
  sort?: Sort;
  limit?: number;
}

export const recordKeys = {
  list: (pid: string, query: RecordQuery) => ["projects", pid, "records", query] as const,
  facets: (pid: string) => ["projects", pid, "records", "facets"] as const,
  detail: (pid: string, rid: string) => ["projects", pid, "records", rid] as const,
};

/** The records table, a page at a time, for the virtualised list. */
export const recordsQuery = (pid: string, query: RecordQuery = {}) =>
  infiniteQueryOptions({
    queryKey: recordKeys.list(pid, query),
    initialPageParam: undefined as string | undefined,
    queryFn: async ({ pageParam, signal }) =>
      unwrap(
        await api.GET("/api/v1/projects/{pid}/records", {
          signal,
          params: {
            path: { pid },
            query: { ...query, limit: query.limit ?? 50, cursor: pageParam },
          },
        }),
      ),
    getNextPageParam: (page: RecordPage) => page.next_cursor ?? undefined,
  });

export const facetsQuery = (pid: string) =>
  queryOptions({
    queryKey: recordKeys.facets(pid),
    queryFn: async ({ signal }) =>
      unwrap(
        await api.GET("/api/v1/projects/{pid}/records/facets", {
          signal,
          params: { path: { pid } },
        }),
      ),
  });

export const recordQuery = (pid: string, rid: string) =>
  queryOptions({
    queryKey: recordKeys.detail(pid, rid),
    queryFn: async ({ signal }) =>
      unwrap(
        await api.GET("/api/v1/projects/{pid}/records/{rid}", {
          signal,
          params: { path: { pid, rid } },
        }),
      ),
  });
